juju/model.py

```diff
@@ -0,0 +1,166 @@
+        # relation data.
+        return hash(self.name)
+
+    def __eq__(self, other):
+        if not isinstance(other, Unit):
+            raise TypeError(f"'==' not supported between instances of '{type(self).__name__}' and '{type(other).__name__}'")
+        return self.name == other.name
+
+
+class RelationMap(dict):
+    """Map of relation names to lists of Relation instances.
+
+    The data for each relation name may be lazy-loaded.
+    """
+    def __init__(self, data):
+        super().__init__()
+        for relation_name, relations in data.items():
+            if relations is not None:
+                relations = [Relation(relation_id, relation_data)
+                             for relation_id, relation_data in sorted(relations.items())]
+            self[relation_name] = relations
+
+    def __getitem__(self, relation_name):
+        value = super().__getitem__(relation_name)
+        if value is None:
+            # Lazy-load the list of relations for this relation name.
+            output = run(['relation-ids', relation_name, '--format=yaml'], stdout=PIPE, check=True).stdout
+            value = [Relation(relation_id) for relation_id in yaml.safe_load(output.decode('utf8'))]
+            self[relation_name] = value
+        return value
+
+
+class Relation:
+    def __init__(self, relation_id, data=None):
+        self.relation_id = relation_id
+        self.data = RelationData(relation_id, data)
+        if data is not None:
+            self._apps = [key for key in self.data if isinstance(key, Application)]
+        else:
+            self._apps = None
+
+    @property
+    def apps(self):
+        if self._apps is None:
+            # Lazy-load the apps on this relation.
+            output = run(['relation-list', '-r', self.relation_id, '--format=yaml'], stdout=PIPE, check=True).stdout
+            member_names = yaml.safe_load(output.decode('utf8'))
+            units_by_app = {}
+            for member_name in sorted(member_names):
+                if '/' not in member_name:
+                    # Handle possible future support from Juju for application relation data.
+                    continue
+                app_name = member_name.split('/')[0]
+                units_by_app.setdefault(app_name, []).append(Unit(member_name))
+            self._apps = [Application(app_name, units) for app_name, units in sorted(units_by_app.items())]
+        return self._apps
+
+
+class RelationData(dict):
+    def __init__(self, relation_id, relation_data=None):
+        super().__init__()
+        self.relation_id = relation_id
+        if relation_data is not None:
+            app_units = {}
+            for app_or_unit_name, member_data in sorted(relation_data.items()):
+                # Apps and units are represented in the same data structure to match the expected
+                # pattern for interacting with Juju and how we present it to the charm code. However,
```

Comment:
The part of the comment above isn't entirely clear. I also don't understand why we're getting such inconsistent data here, which is certainly related to me not understanding the comment justifying this.

It would seem that not having those two mixed here would be possible and good.